sarc/users/db.py

```diff
@@ -0,0 +1,303 @@
+"""
+This script is basically a wrapper around the "read_mila_ldap.py" script.
+Instead of taking arguments from the command line, it takes them from
+the SARC configuration file.
+
+This is possible because the "read_mila_ldap.py" script has a `run` function
+that takes the arguments as parameters, so the argparse step comes earlier.
+
+As a result of running this script, the values in the collection
+referenced by "cfg.ldap.mongo_collection_name" will be updated.
+"""
+
+import logging
+
+from pydantic import UUID4, field_serializer
+from pydantic_mongo import AbstractRepository, PydanticObjectId
+
+from sarc.config import config
+from sarc.core.models.users import MemberType, UserData
+from sarc.core.models.validators import DateOverlapError, ValidField, ValidTag
+from sarc.core.scraping.users import MatchID, UserMatch
+
+logger = logging.getLogger(__name__)
+
+
+class UserDB(UserData):
+    """
+    Data for one user
+    """
+
+    # Database ID
+    id: PydanticObjectId | None = None
+
+    @field_serializer("member_type")
+    def _valid_enum_to_str(self, value: ValidField[MemberType]) -> ValidField[str]:
+        return ValidField(
+            values=[
+                ValidTag(
+                    value=vt.value.value,
+                    valid_start=vt.valid_start,
+                    valid_end=vt.valid_end,
+                )
+                for vt in value.values
+            ]
+        )
+
+    @field_serializer("co_supervisors")
+    def _serialize_as_list(
+        self, value: ValidField[set[UUID4]]
+    ) -> ValidField[list[UUID4]]:
+        return ValidField(
+            values=[
+                ValidTag(
+                    value=list(v.value),
+                    valid_start=v.valid_start,
+                    valid_end=v.valid_end,
+                )
+                for v in value.values
+            ]
+        )
+
+
+class UserRepository(AbstractRepository[UserDB]):
+    class Meta:
+        collection_name = "users"
+
+    def update_user(self, user: UserMatch) -> None:
```

Comment:
Since user structure has changed, I think we should also update `sarc db init`, especially collection indices in `create_users_indices()` function.